File: noorbedtime-app/scripts/upload-illustrations.py
#!/usr/bin/env python3
"""
Upload all WebP illustrations from images-compressed/ to Supabase Storage.
Bucket: illustrations (must be public)

Requirements: pip3 install httpx
Usage: python3 scripts/upload-illustrations.py
"""
import os
import sys
import time
from pathlib import Path

try:
    import httpx
except ImportError:
    print("Installing httpx...")
    os.system("pip3 install httpx -q")
    import httpx

SUPABASE_URL = "https://unlaqymnoapznqxeqgom.supabase.co"
SUPABASE_KEY = os.environ.get("SUPABASE_SERVICE_KEY") or os.environ.get(
    "NEXT_PUBLIC_SUPABASE_ANON_KEY",
    "sb_publishable_eLMd3ATdRXTrSNUl9ovIwg_uyC1Fnkx"
)
BUCKET = "illustrations"
IMAGES_DIR = Path(__file__).parent.parent.parent / "images-compressed"

HEADERS = {
    "apikey": SUPABASE_KEY,
    "Authorization": f"Bearer {SUPABASE_KEY}",
}

def ensure_bucket(client: httpx.Client):
    resp = client.get(f"{SUPABASE_URL}/storage/v1/bucket/{BUCKET}", headers=HEADERS)
    if resp.status_code == 404:
        print(f"Creating bucket '{BUCKET}'...")
        r = client.post(
            f"{SUPABASE_URL}/storage/v1/bucket",
            headers=HEADERS,
            json={"id": BUCKET, "name": BUCKET, "public": True},
        )
        if r.status_code not in (200, 201):
            print(f"  ✗ Could not create bucket: {r.text}")
            print("  → Create bucket manually in Supabase Dashboard > Storage")
            return False
        print(f"  ✓ Bucket '{BUCKET}' created (public)")
    elif resp.status_code == 200:
        print(f"  ✓ Bucket '{BUCKET}' already exists")
    else:
        print(f"  ⚠ Bucket check returned {resp.status_code}: {resp.text[:200]}")
        print("  → Ensure bucket exists in Supabase Dashboard > Storage")
    return True

def upload_file(client: httpx.Client, local_path: Path, storage_path: str) -> bool:
    with open(local_path, "rb") as f:
        data = f.read()

    upload_headers = {
        **HEADERS,
        "Content-Type": "image/webp",
        "x-upsert": "true",
    }
    resp = client.post(
        f"{SUPABASE_URL}/storage/v1/object/{BUCKET}/{storage_path}",
        headers=upload_headers,
        content=data,
        timeout=30,
    )
    return resp.status_code in (200, 201)

def main():
    webp_files = sorted(IMAGES_DIR.rglob("*.webp"))
    if not webp_files:
        print(f"No WebP files found in {IMAGES_DIR}")
        sys.exit(1)

    print(f"Found {len(webp_files)} WebP files in {IMAGES_DIR}")
    print(f"Uploading to Supabase Storage bucket '{BUCKET}'...\n")

    with httpx.Client() as client:
        bucket_ok = ensure_bucket(client)
        if not bucket_ok:
            print("\n⚠ Proceeding anyway — uploads may fail without bucket.")

        success = 0
        failed = []

        for i, path in enumerate(webp_files, 1):
            # storage path: story-slug/page-00.webp
            relative = path.relative_to(IMAGES_DIR)
            storage_path = str(relative).replace("\\", "/")

            ok = upload_file(client, path, storage_path)
            if ok:
                success += 1
                if i % 50 == 0 or i == len(webp_files):
                    print(f"  [{i}/{len(webp_files)}] ✓ {success} uploaded, {len(failed)} failed")
            else:
                failed.append(storage_path)
                print(f"  [{i}/{len(webp_files)}] ✗ Failed: {storage_path}")

            # Small delay every 100 files to avoid rate limits
            if i % 100 == 0:
                time.sleep(1)

    print(f"\n{'='*50}")
    print(f"✓ {success}/{len(webp_files)} illustrations uploaded")
    if failed:
        print(f"✗ {len(failed)} failed:")
        for f in failed[:10]:
            print(f"  - {f}")
        if len(failed) > 10:
            print(f"  ... and {len(failed)-10} more")
    else:
        print("All illustrations uploaded successfully!")
    print(f"\nPublic URL base:")
    print(f"  {SUPABASE_URL}/storage/v1/object/public/{BUCKET}/{{slug}}/page-00.webp")

if __name__ == "__main__":
    main()

File: noorbedtime-app/scripts/upload-illustrations.mjs
#!/usr/bin/env node
/**
 * Upload all WebP illustrations to Supabase Storage.
 * Run AFTER creating the 'illustrations' bucket in Supabase Dashboard.
 *
 * Usage:
 *   SUPABASE_SERVICE_KEY=your_service_role_key node scripts/upload-illustrations.mjs
 *
 * Or set it in .env.local as SUPABASE_SERVICE_KEY=...
 */

import { createClient } from "@supabase/supabase-js";
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// Load .env.local manually
const envPath = path.join(__dirname, "../.env.local");
if (fs.existsSync(envPath)) {
  for (const line of fs.readFileSync(envPath, "utf-8").split("\n")) {
    const [k, ...v] = line.split("=");
    if (k && !k.startsWith("#") && v.length) {
      process.env[k.trim()] = v.join("=").trim();
    }
  }
}

const SUPABASE_URL = process.env.NEXT_PUBLIC_SUPABASE_URL;
const SERVICE_KEY =
  process.env.SUPABASE_SERVICE_KEY || process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;
const BUCKET = "illustrations";
const IMAGES_DIR = path.join(__dirname, "../../images-compressed");

if (!SUPABASE_URL || !SERVICE_KEY) {
  console.error("Missing NEXT_PUBLIC_SUPABASE_URL or SUPABASE_SERVICE_KEY");
  process.exit(1);
}

const supabase = createClient(SUPABASE_URL, SERVICE_KEY);

function walkDir(dir) {
  const files = [];
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) files.push(...walkDir(full));
    else if (entry.name.endsWith(".webp")) files.push(full);
  }
  return files;
}

async function ensureBucket() {
  const { data, error } = await supabase.storage.getBucket(BUCKET);
  if (error?.message?.includes("not found") || !data) {
    console.log(`Creating bucket '${BUCKET}'...`);
    const { error: createErr } = await supabase.storage.createBucket(BUCKET, {
      public: true,
      allowedMimeTypes: ["image/webp"],
    });
    if (createErr) {
      console.error(`✗ Could not create bucket: ${createErr.message}`);
      console.log(
        "\n→ Please create bucket manually in Supabase Dashboard > Storage:\n" +
          "  Name: illustrations\n  Public: ON\n"
      );
      return false;
    }
    console.log(`✓ Bucket '${BUCKET}' created`);
  } else {
    console.log(`✓ Bucket '${BUCKET}' already exists`);
  }
  return true;
}

async function main() {
  if (!fs.existsSync(IMAGES_DIR)) {
    console.error(`Images directory not found: ${IMAGES_DIR}`);
    process.exit(1);
  }

  const files = walkDir(IMAGES_DIR);
  console.log(`Found ${files.length} WebP files\n`);

  await ensureBucket();

  let success = 0;
  const failed = [];
  const CONCURRENCY = 5;

  async function uploadFile(filePath) {
    const storagePath = path.relative(IMAGES_DIR, filePath).replace(/\\/g, "/");
    const buffer = fs.readFileSync(filePath);

    const { error } = await supabase.storage
      .from(BUCKET)
      .upload(storagePath, buffer, {
        contentType: "image/webp",
        upsert: true,
      });

    if (error) {
      failed.push({ path: storagePath, error: error.message });
      return false;
    }
    return true;
  }

  // Upload in batches
  for (let i = 0; i < files.length; i += CONCURRENCY) {
    const batch = files.slice(i, i + CONCURRENCY);
    const results = await Promise.all(batch.map(uploadFile));
    success += results.filter(Boolean).length;

    if ((i + CONCURRENCY) % 50 === 0 || i + CONCURRENCY >= files.length) {
      const done = Math.min(i + CONCURRENCY, files.length);
      process.stdout.write(
        `\r  [${done}/${files.length}] ✓ ${success} uploaded, ✗ ${failed.length} failed`
      );
    }
  }

  console.log(`\n\n${"=".repeat(50)}`);
  console.log(`✓ ${success}/${files.length} illustrations uploaded to Supabase Storage`);

  if (failed.length > 0) {
    console.log(`\n✗ Failed (${failed.length}):`);
    failed.slice(0, 10).forEach((f) => console.log(`  - ${f.path}: ${f.error}`));
    if (failed.length > 10) console.log(`  ... and ${failed.length - 10} more`);
  }

  console.log(`\nPublic URL format:`);
  console.log(`  ${SUPABASE_URL}/storage/v1/object/public/${BUCKET}/{slug}/page-00.webp`);
}

main().catch(console.error);
